### PR TITLE
Feature/carma messenger ambassador

### PR DESCRIPTION
**Pull request by EricChen-Lei**
# PR Details
## Description

Implement carma messenger ambassador
## Related Jira Key
CXC-88

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**CONTRIBUTING**](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: .vscode/settings.json
Comment (chengyuan0124) [inline .vscode/settings.json:1]: Is this your PC's config? if yes then might need to be removed

File: co-simulation/fed/mosaic-carma-messenger/src/main/java/org/eclipse/mosaic/fed/carmamessenger/ambassador/CarmaMessengerInstanceManager.java
Comment (chengyuan0124) [inline co-simulation/fed/mosaic-carma-messenger/src/main/java/org/eclipse/mosaic/fed/carmamessenger/ambassador/CarmaMessengerInstanceManager.java:45]: 5374 might be occupied by the NS-3 adapter which connects CARMA Platform and CARMA Ambassador.

File: co-simulation/fed/mosaic-carma-messenger/src/main/java/org/eclipse/mosaic/fed/carmamessenger/ambassador/CarmaMessengerRegistrationMessage.java
Comment (chengyuan0124) [inline co-simulation/fed/mosaic-carma-messenger/src/main/java/org/eclipse/mosaic/fed/carmamessenger/ambassador/CarmaMessengerRegistrationMessage.java:33]: Do we receive the messengerEmergencyState when the CARMA Messenger sends its registration to the CARMA Messenger Ambassador?

File: .vscode/settings.json
Comment (EricChen-Lei) [reply .vscode/settings.json:1]: removed

File: co-simulation/fed/mosaic-carma-messenger/src/main/java/org/eclipse/mosaic/fed/carmamessenger/ambassador/CarmaMessengerInstanceManager.java
Comment (EricChen-Lei) [reply co-simulation/fed/mosaic-carma-messenger/src/main/java/org/eclipse/mosaic/fed/carmamessenger/ambassador/CarmaMessengerInstanceManager.java]: changed port

File: co-simulation/fed/mosaic-carma-messenger/src/main/java/org/eclipse/mosaic/fed/carmamessenger/ambassador/CarmaMessengerRegistrationReceiver.java
Comment (EricChen-Lei) [reply co-simulation/fed/mosaic-carma-messenger/src/main/java/org/eclipse/mosaic/fed/carmamessenger/ambassador/CarmaMessengerRegistrationReceiver.java]: changed port

File: co-simulation/fed/mosaic-carma-messenger/src/main/java/org/eclipse/mosaic/fed/carmamessenger/ambassador/CarmaMessengerRegistrationMessage.java
Comment (EricChen-Lei) [reply co-simulation/fed/mosaic-carma-messenger/src/main/java/org/eclipse/mosaic/fed/carmamessenger/ambassador/CarmaMessengerRegistrationMessage.java]: changed name

File: co-simulation/fed/mosaic-carma-messenger/src/test/java/org/eclipse/mosaic/fed/carmamessenger/ambassador/CarmaMessengerInstanceManagerTest.java
Comment (chengyuan0124) [inline co-simulation/fed/mosaic-carma-messenger/src/test/java/org/eclipse/mosaic/fed/carmamessenger/ambassador/CarmaMessengerInstanceManagerTest.java]: where is this function define and do we need that for CARMA Messenger vehicle

File: co-simulation/fed/mosaic-carma-messenger/src/test/java/org/eclipse/mosaic/fed/carmamessenger/ambassador/CarmaMessengerInstanceManagerTest.java
Comment (chengyuan0124) [inline co-simulation/fed/mosaic-carma-messenger/src/test/java/org/eclipse/mosaic/fed/carmamessenger/ambassador/CarmaMessengerInstanceManagerTest.java]: where is this function define and do we need that for CARMA Messenger vehicle

File: co-simulation/fed/mosaic-carma-messenger/src/main/java/org/eclipse/mosaic/fed/carmamessenger/ambassador/CarmaMessengerInstance.java
Comment (chengyuan0124) [inline co-simulation/fed/mosaic-carma-messenger/src/main/java/org/eclipse/mosaic/fed/carmamessenger/ambassador/CarmaMessengerInstance.java:115]: Do we still use the function name same with CARMA Ambassador or we can change to something like "carmaMessengerVehicleId"?

File: co-simulation/fed/mosaic-carma-messenger/src/main/java/org/eclipse/mosaic/fed/carmamessenger/ambassador/CarmaMessengerInstance.java
Comment (chengyuan0124) [inline co-simulation/fed/mosaic-carma-messenger/src/main/java/org/eclipse/mosaic/fed/carmamessenger/ambassador/CarmaMessengerInstance.java:119]: same with above

**Comment by kjrush** [comment]
I'm a bit concerned about the amount of duplication in this one, it looks like a lot of this is the same as the CARMA Platform ambassador, which is fine and expected. But rather than simply duplicating the code, we should be looking at how to refactor it to minimize duplication. Can we inherit directly from the CARMAPlatformAmbassador class and override the specific functions we need to change to support messenger? Are there parts of the CARMA Platform Ambassador that should be refactored out into a base-class that both Platform and Messenger ambassadors inherit from? Should some of the existing classes just be renamed and reused as-is instead of duplicated?

File: co-simulation/fed/mosaic-carma-messenger/src/main/java/org/eclipse/mosaic/fed/carmamessenger/ambassador/CarmaMessengerInstance.java
Comment (chengyuan0124) [inline co-simulation/fed/mosaic-carma-messenger/src/main/java/org/eclipse/mosaic/fed/carmamessenger/ambassador/CarmaMessengerInstance.java]: It might be better to change to getSumoRoleName since messenger vehicle is generated in SUMO

File: co-simulation/fed/mosaic-carma-messenger/src/main/java/org/eclipse/mosaic/fed/carmamessenger/ambassador/CarmaMessengerInstance.java
Comment (chengyuan0124) [inline co-simulation/fed/mosaic-carma-messenger/src/main/java/org/eclipse/mosaic/fed/carmamessenger/ambassador/CarmaMessengerInstance.java]: same with above

File: co-simulation/fed/mosaic-carma-messenger/src/main/java/org/eclipse/mosaic/fed/carmamessenger/ambassador/CarmaMessengerInstanceManager.java
Comment (chengyuan0124) [inline co-simulation/fed/mosaic-carma-messenger/src/main/java/org/eclipse/mosaic/fed/carmamessenger/ambassador/CarmaMessengerInstanceManager.java]: log should be changed to CARMA messenger to avoid misunderstanding

File: co-simulation/fed/mosaic-carma-messenger/src/main/java/org/eclipse/mosaic/fed/carmamessenger/ambassador/CarmaMessengerInstanceManager.java
Comment (chengyuan0124) [inline co-simulation/fed/mosaic-carma-messenger/src/main/java/org/eclipse/mosaic/fed/carmamessenger/ambassador/CarmaMessengerInstanceManager.java:181]: same with above

File: co-simulation/lib/mosaic-common-utils/src/main/java/org/eclipse/mosaic/lib/CommonUtil/ambassador/CommonMessageAmbassador.java
Comment (kjrush) [inline co-simulation/lib/mosaic-common-utils/src/main/java/org/eclipse/mosaic/lib/CommonUtil/ambassador/CommonMessageAmbassador.java:150]: At runtime, will the formatting for this logger show the actual inheriting class or will all log messages now seem to come form "CommonMessageAmbassador"? Might be confusing to wade through if all the CARMA ambassador's log statements get mixed together, worth considering if we can add the subclass name to these log statements somehow. There's got to be a way. My java's a bit rusty so I'm not sure what `this.getClass()` would return here, whether that'd give the superclass or the subclass, but something along those lines has to work.

**Review by kjrush — APPROVED** [review]
Generally looks good to me! I've got a few minor nitpicks that you can choose to take a look at or ignore, but I think this is a good approach overall!

File: co-simulation/fed/mosaic-carma-messenger/src/main/java/org/eclipse/mosaic/fed/carmamessenger/ambassador/CarmaMessengerInstanceManager.java
Comment (EricChen-Lei) [reply co-simulation/fed/mosaic-carma-messenger/src/main/java/org/eclipse/mosaic/fed/carmamessenger/ambassador/CarmaMessengerInstanceManager.java:33]: Thank you Kyle, I must have missed this part, will delete it right away

File: co-simulation/lib/mosaic-common-utils/src/main/java/org/eclipse/mosaic/lib/CommonUtil/ambassador/CommonMessageAmbassador.java
Comment (EricChen-Lei) [reply co-simulation/lib/mosaic-common-utils/src/main/java/org/eclipse/mosaic/lib/CommonUtil/ambassador/CommonMessageAmbassador.java:150]: I actually tested with this.getClass(), the only problem I potentially see it's maybe it too long and complicate for a log, so I use this.getClass().getSimpleName() instead, it ouputs the class name but not the package name.

Comment (kjrush) [reply co-simulation/lib/mosaic-common-utils/src/main/java/org/eclipse/mosaic/lib/CommonUtil/ambassador/CommonMessageAmbassador.java]: Yeah, I remember that method now. getSimpleName() should work great for what we're looking for. I just wasn't sure whether it'd return "CommonMessengerAmbassador" or the name of the actual subclass, but if it does actually return the  subclass name then that's exactly what we'd want!

**Review by kjrush — APPROVED** [review]
I think these changes look good to me, there might be some further cleanup that could be done w.r.t. the class names being used in logging, but I don't think that's worth holding this merge up. If you want to add the class name more broadly to the log statements as a prefix "<CLASS_NAME>: New instance registered of type ...." etc. then please do so, but I'll leave that up to you. I'll approve for now so you can merge whenever you're ready.

**Pull request merged**
